docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,430 @@
+---
+title: Runtime Hooks for Add-on Management
+authors:
+  - "@killianmuldoon"
+  - "@ykakarap"
+reviewers:
+  - "@vincepri"
+  - "@CecileRobertMichon"
+  - "@enxebre"
+  - "@fabriziopandini"
+  - "@sbueringer"
+creation-date: 2022-04-14
+last-updated: 2022-04-14
+status: implementable
+replaces:
+see-also:
+superseded-by:
+---
+
+# Runtime hooks for Add-on Management
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Glossary](#glossary)
+* [Summary](#summary)
+* [Motivation](#motivation)
+	* [Goals](#goals)
+	* [Non-goals](#non-goals)
+    * [Future work](#futurework)
+* [Proposal](#proposal)
+    * [User Stories](#user-stories)
+    * [Runtime hook definitions](#runtime-hook-definitions)
+      * [Before Cluster Create](#before-cluster-create)
+      * [After Control Plane Initialized](#after-control-plane-initialized)
+      * [Before Cluster Upgrade](#before-cluster-upgrade)
+      * [After Control Plane Upgrade](#after-control-plane-upgrade)
+      * [After Cluster Upgrade](#after-cluster-upgrade)
+      * [Before Cluster Delete](#before-cluster-delete)
+    * [Runtime Extensions developer guide](#runtime-extensions-developer-guide)
+    * [Security Model](#security-model)
+    * [Risks and Mitigations](#risks-and-mitigations)
+      * [Runtime Extension blocks Cluster lifecycle indefinitely](#runtime-extension-blocks-cluster-lifecycle-indefinitely)
+* [Alternatives](#alternatives)
+  * [External components watching CAPI resources without hooks](#external-components-watching-capi-resources-without-hooks)
+  * [OpenAPI spec implementation alternatives](#openapi-spec-implementation-alternatives)
+    * [Adding Cluster info to request vs providing only the Cluster name](#adding-cluster-info-to-request-vs-providing-only-the-cluster-name)
+    * [Embedding CAPI types in request vs using runtime.RawExtension](#embedding-capi-types-in-request-vs-using-runtimerawextension)
+* [Upgrade Strategy](#upgrade-strategy)
+    * [Cluster API version upgrade](#clusterapi-version-upgrade)
+    * [Kubernetes version upgrade](#kubernetes-version-upgrade)
+* [Additional Details](#additional-details)
+    * [Test Plan](#test-plan)
+    * [Graduation Criteria](#graduation-criteria)
+    * [Version Skew Strategy](#version-skew-strategy)
+* [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+
+## Summary
+
+This proposal introduces a set of Runtime Hooks designed for allowing the implementation of add-on orchestration solutions on top of Cluster API.
```

Comment:
I'm proposing to expand the summary a little bit in order to address @CecileRobertMichon comment

>  ... it's unclear to me from reading this proposal what the relationship is between this and the Cluster Addons proposal which also seems to be moving forward. I also think we should make it more clear that these hooks don't have to be used for addon management, they could be leveraged for anything, even if the primary motivation of the proposal is addon management. 

```suggestion
This proposal introduces a set of Runtime Hooks designed for providing the foundation for the implementation of add-on orchestration solutions on top of Cluster API.

However, given that the hooks defined in this proposal are going to model common events in the Cluster lifecycle they could be used for other use cases beyond add-on management, but for the sake of having a focused and actionable scope, we are not exploring those option in this document. 
```

I would also take an AI to work with @jackfrancis @Jont828 to clarify in their document how the Cluster Addons provider for helm could take care of the Hooks introduced in this document